emma-language/src/main/scala/org/emmalanguage/api/DataBag.scala

```diff
@@ -311,6 +316,34 @@ trait DataBag[A] extends Serializable {
           }
       })._1
   }
+
+  // -----------------------------------------------------
+  // equals, hashCode and toString
+  // -----------------------------------------------------
+
+  override def equals(o: Any) = o match {
+    case that: DataBag[A] => {
+      val thisVals = fetch()
+      val thatVals = that.fetch()
+      (thisVals.size == thatVals.size) &&
+        (thisVals.toSet diff thatVals.toSet).isEmpty
+    }
+    case _ => false
+  }
+
+  override def hashCode(): Int =
+    scala.util.hashing.MurmurHash3.unorderedHash(fetch())
+
+  override def toString: String =
+    fetch().toString
```

Comment:
This a user-facing API, we should be very careful what we expose. For example, if you evaluate a `DataBag` in the Scala REPL, per default the result is rendered using a `toString()` call. 

I suggest to implement this by delegating to `rep.toString`.